### PR TITLE
Add routines to compute ghosted subdomain and interface elements

### PR DESCRIPTION
**Pull request by jbadger95**
`get_ghost_subd` computes:

- Ghosted subdomain (defined to be any element that shares any _modified_ element node with a rank's subdomain elements)
- Interface elements (defined to be any element with a  _modified_ element on the interface between the rank's subdomain and any other rank's subdomain; interface elements are a subset ghosted subdomain)

This routine is fast and scalable; it only performs subdomain-local operations during the computation. Simple serial computation of the ghosted subdomain and interface elements is also implemented for verification in DEBUG_MODE (simple routines are ~100-1000x slower on large meshes).

You may want to move these routines; I didn't put them inside the `par_mesh` module because I don't want to create a circular dependence if called in `update_elem_order`

### CODE REVIEW

**Comment by jbadger95** [comment]
It took a while to come up with the scalable version of this routine; it turns out it's not trivial to compute the ghosted subdomain when each rank only has information about its subdomain. Sorry if it's hard to follow but any problems become immediately known when run in `DEBUG_MODE` due to the verification routines so it shouldn't be too dangerous to merge.

**Comment by stefanhenneking** [comment]
This PR adds features that are used only by the MG solver?

**Comment by jbadger95** [comment]
`enforce_min_rule` uses the ghosted subdomain. It's mostly used by my solver but it will also be used for distributing the datastructure. I also think it's generally useful; subdomains usually don't need to operate on more than their ghosted subdomain (and shouldn't for scalability).

File: trunk/src/modules/par_mesh.F90 [hunk 1]
```diff
@@ -41,14 +41,14 @@ module par_mesh
...
-   integer :: ierr
+   integer :: ierr, base, remainder
    integer :: tag, count, src, dest
    VTYPE, allocatable :: buf(:)
 !
 !..auxiliary variables
    integer :: subd_next(NRELES), nodm(MAXNODM), nodesl(27)
    integer :: iel, inod, iproc, mdle, nod, subd, subd_size
-   integer :: nrnodm, nrdof_nod
+   integer :: nrnodm, nrdof_nod, proc, el
```
Comment (stefanhenneking) [inline trunk/src/modules/par_mesh.F90:51]: there's already `iproc` and `iel` that can be used
```suggestion
```

File: trunk/src/modules/par_mesh.F90 [hunk 2]
```diff
@@ -57,16 +57,28 @@ subroutine distr_mesh()
...
    if ((ZOLTAN_LB .eq. 0) .or. (.not. DISTRIBUTED)) then
-      subd_size = (NRELES+NUM_PROCS-1)/NUM_PROCS
-      iproc=0
-      do iel=1,NRELES
-!     ...decide subdomain for iel
-         subd_next(iel) = iproc
-         iproc = iel/subd_size
+!
+      base = NRELES/NUM_PROCS
+      remainder = mod(NRELES,NUM_PROCS)
```
Comment (stefanhenneking) [inline trunk/src/modules/par_mesh.F90:68]: Thanks for fixing this partitioning.

File: trunk/src/modules/par_mesh.F90 [hunk 3]
```diff
@@ -57,16 +57,28 @@ subroutine distr_mesh()
...
+!
+!  ...first remainder procs get base+1, rest get base
+      do proc = 0,NUM_PROCS-1
+         if (proc .lt. remainder) then
+            subd_size = base + 1
+         else
+            subd_size = base
+         endif
+         subd_next(el+1:el+subd_size) = proc
+         el = el + subd_size
```
Comment (stefanhenneking) [inline trunk/src/modules/par_mesh.F90:81]:
```suggestion
!  ...element offset
      iel=0
!
!  ...first remainder procs get base+1, rest get base
      do iproc = 0,NUM_PROCS-1
         if (iproc .lt. remainder) then
            subd_size = base + 1
         else
            subd_size = base
         endif
         subd_next(iel+1:iel+subd_size) = iproc
         iel = iel + subd_size
```

File: trunk/src/modules/par_mesh.F90 [hunk 4]
```diff
@@ -129,7 +141,7 @@ subroutine distr_mesh()
                write(6,130) '[', RANK, ']: ',   &
                   'Receiving data from [',subd,'], nod = ',nod
             endif
-            count = nrdof_nod; src = subd; tag = nod
+            count = nrdof_nod; src = subd; tag = mod(nod,200000)
```
Comment (stefanhenneking) [inline trunk/src/modules/par_mesh.F90:144]: This is probably appearing in some other parts of the code. How large are the tags allowed to be? Integer limit?

**Review by stefanhenneking — CHANGES_REQUESTED** [review]
I'd prefer to put definitions and routines related to ghosted subdomain and interface in a new, separate module.

File: trunk/src/modules/par_mesh.F90 (hunk 4 above)
Comment (jbadger95) [reply trunk/src/modules/par_mesh.F90:144]: It was in `update_gdof` and `update_Ddof` as well; I have it updated throughout my code but I'm not sure they've all made it in. It's only guaranteed to be a bit of over 200k (2^18) if I remember correctly. The limit used to be 2 million for Stampede2 but on Frontera the limit appears to be the minimum 2^18.

**Comment by jbadger95** [comment]
That's fine; I actually did have it in a separate module (under `src/mpi/get_ghost_subd`) but I thought last time I told you about it you said you wanted it in`par_mesh` (maybe I'm remembering wrong). Anyway, I don't mind. I think we should keep the element orderings in the global datastructure but it doesn't particularly matter where we put the subroutines

File: trunk/src/modules/par_mesh.F90 (hunk 4 above)
Comment (stefanhenneking) [reply trunk/src/modules/par_mesh.F90:144]: Intel states `If a code exceeds the tag value of 32767 or 15 bits, the actual upper limit must be queried in the application. Otherwise, it creates a portability risk.`
Should we introduce a max tag parameter in the mpi_wrapper module, set it to a currently safe number, and use it throughout?

**Comment by stefanhenneking** [comment]
As a module name, I think `par_ghost` or `ghost_subd` would be preferable to `get_ghost_subd` (which sounds more like a function name returning a value)

**Comment by jbadger95** [comment]
Just wondering why they need to be put in a module (I misspoke earlier, i meant they were in a file)? I'm fine with `par_ghost`, sounds alright but I feel like a module just makes extra work to call it if we aren't actually storing persistent variables with it.

**Comment by stefanhenneking** [comment]
I'm thinking that defining scopes is generally preferable to having subroutines defined with global scope as is done for many routines in hp3D. It'd be generally much better (imo) to have modules that define interfaces for the subroutines use everywhere.

**Comment by jbadger95** [comment]
Okay; sounds good, I'll move it to a module `par_ghost`

File: trunk/src/modules/par_ghost.F90 [hunk 5]
```diff
@@ -0,0 +1,683 @@
...
+!                ELEM_GHOST : ELEM_SUBD plus elements touching my subd
+!                ELEM_INTERF: mdle nodes touching subd interface
+!
+!                both are stored in natural order of elements
+!
+!-----------------------------------------------------------------------
+module par_ghost
+
+
+contains
```
Comment (stefanhenneking) [inline trunk/src/modules/par_ghost.F90:22]:
```suggestion
implicit none
!
contains
```

File: trunk/src/modules/par_mesh.F90 [hunk 6]
```diff
@@ -57,16 +57,28 @@ subroutine distr_mesh()
       write(6,100) 'start distr_mesh, DISTRIBUTED = ', DISTRIBUTED
    endif
    100 format(A,L2)
+!
+   if (NUM_PROCS.eq.1) return
 !
 !..1. Determine new partition
    call MPI_BARRIER (MPI_COMM_WORLD, ierr); start_time = MPI_Wtime()
    if ((ZOLTAN_LB .eq. 0) .or. (.not. DISTRIBUTED)) then
```
Comment (stefanhenneking) [inline trunk/src/modules/par_mesh.F90:65]: I just noticed now that we have enabled the `hp3gen` routine to distribute the mesh rightaway, this line prevents the user from setting a partitioner other than `ZOLTAN_LB=0` for initial distribution.

I think we can change it simply to:

`if (ZOLTAN_LB .eq. 0) then`

**Comment by stefanhenneking** [comment]
@jbadger95 can you check on the comment/suggestions. then, it's ready to be merged imo.

File: trunk/src/modules/par_mesh.F90 (hunk 6 above)
Comment (jbadger95) [reply trunk/src/modules/par_mesh.F90:65]: I don't think it should be changed. Some partitioners (e.g. weighted graph partitioner) require options to be set to enable starting with a non-distributed mesh. In fact many of the Zoltan partitioners return errors if there are fewer elements than processors and it's fairly common to start with single element meshes and then refine before partitioning (which you would do after `hp3gen`). Partitioning first using the simple method speeds up initial calls to `update_gdof` and `update_Ddof`, the user can repartition how they wish after that.

File: trunk/src/modules/par_mesh.F90 (hunk 4 above)
Comment (jbadger95) [reply trunk/src/modules/par_mesh.F90:144]: I'll do this in the wrapper changes I have planned.

Comment (stefanhenneking) [reply trunk/src/modules/par_mesh.F90:144]: Sounds good!

File: trunk/src/modules/par_mesh.F90 (hunk 6 above)
Comment (stefanhenneking) [reply trunk/src/modules/par_mesh.F90:65]: I'm ok with that; in this case though, we should remove the option in problem's `set_initial_mesh` routines for the user to set the Load Balancer before `hp3gen` is called (because the choice will be simply ignored here).

**Review by stefanhenneking — APPROVED** [review]
This is good to @jbadger95 (note I did not review any of the code in the new module `par_ghost` since it doesn't affect existing routines.
